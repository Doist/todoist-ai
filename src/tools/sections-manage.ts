import type { Section } from '@doist/todoist-api-typescript'
import { z } from 'zod'
import { getToolOutput } from '../mcp-helpers.js'
import type { TodoistTool } from '../todoist-tool.js'
import { formatNextSteps } from '../utils/response-builders.js'
import { ToolNames } from '../utils/tool-names.js'

const { TASKS_ADD_MULTIPLE, TASKS_LIST_FOR_CONTAINER, OVERVIEW, SECTIONS_SEARCH } = ToolNames

const ArgsSchema = {
    id: z
        .string()
        .min(1)
        .optional()
        .describe(
            'The ID of the section to update. If provided, updates the section. If omitted, creates a new section.',
        ),
    name: z.string().min(1).describe('The name of the section.'),
    projectId: z
        .string()
        .min(1)
        .optional()
        .describe(
            'The ID of the project to add the section to. Required when creating a new section (when id is not provided).',
        ),
}

const sectionsManage = {
    name: ToolNames.SECTIONS_MANAGE,
    description:
        'Add a new section to a project or update an existing section. If id is provided, updates the section. If id is omitted, creates a new section (requires projectId).',
    parameters: ArgsSchema,
    async execute(args, client) {
        if (args.id) {
            // Update existing section
            const section = await client.updateSection(args.id, { name: args.name })
            const textContent = generateTextContent({ section, operation: 'updated' })

            return getToolOutput({
                textContent,
                structuredContent: {
                    section,
                    operation: 'updated',
                },
            })
        }

        // Create new section - projectId is required
        if (!args.projectId) {
            throw new Error(
                'Error: projectId is required when creating a new section (when id is not provided).',
            )
        }

        const section = await client.addSection({
            name: args.name,
            projectId: args.projectId,
        })

        const textContent = generateTextContent({ section, operation: 'created' })

        return getToolOutput({
            textContent,
            structuredContent: {
                section,
                operation: 'created',
            },
        })
    },
} satisfies TodoistTool<typeof ArgsSchema>

function generateTextContent({
    section,
    operation,
}: {
    section: Section
    operation: 'created' | 'updated'
}): string {
    const action = operation === 'created' ? 'Created' : 'Updated'
    const summary = `${action} section: ${section.name} â€¢ id=${section.id}`

    // Workflow-aware next steps based on operation and section context
    const nextSteps: string[] = []

    if (operation === 'created') {
        // For new sections, suggest immediate setup workflow
        nextSteps.push(
            `Use ${TASKS_ADD_MULTIPLE} with sectionId=${section.id} to add your first tasks`,
        )
        nextSteps.push(
            `Use ${TASKS_LIST_FOR_CONTAINER} with type=section and id=${section.id} to verify setup`,
        )

        // Suggest related organization
        if (section.projectId) {
            nextSteps.push(
                `Use ${OVERVIEW} with projectId=${section.projectId} to see project organization`,
            )
        }
    } else {
        // For updated sections, suggest review and management
        nextSteps.push(
            `Use ${TASKS_LIST_FOR_CONTAINER} with type=section and id=${section.id} to see existing tasks`,
        )
        nextSteps.push(`Use ${SECTIONS_SEARCH} to see all sections in this project`)

        // Suggest task updates if section was renamed
        nextSteps.push('Consider updating task descriptions if section purpose changed')
    }

    const next = formatNextSteps(nextSteps)
    return `${summary}\n${next}`
}

export { sectionsManage }
