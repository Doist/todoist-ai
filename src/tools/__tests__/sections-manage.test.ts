import type { Section, TodoistApi } from '@doist/todoist-api-typescript'
import { jest } from '@jest/globals'
import { sectionsManage } from '../sections-manage.js'
import {
    TEST_IDS,
    createMockSection,
    extractStructuredContent,
    extractTextContent,
} from '../test-helpers.js'

// Mock the Todoist API
const mockTodoistApi = {
    addSection: jest.fn(),
    updateSection: jest.fn(),
} as unknown as jest.Mocked<TodoistApi>

describe('sections-manage tool', () => {
    beforeEach(() => {
        jest.clearAllMocks()
    })

    describe('creating a new section', () => {
        it('should create a section and return result', async () => {
            const mockApiResponse = createMockSection({
                id: TEST_IDS.SECTION_1,
                projectId: TEST_IDS.PROJECT_TEST,
                name: 'test-abc123def456-section',
            })

            mockTodoistApi.addSection.mockResolvedValue(mockApiResponse)

            const result = await sectionsManage.execute(
                { name: 'test-abc123def456-section', projectId: TEST_IDS.PROJECT_TEST },
                mockTodoistApi,
            )

            expect(mockTodoistApi.addSection).toHaveBeenCalledWith({
                name: 'test-abc123def456-section',
                projectId: TEST_IDS.PROJECT_TEST,
            })

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Created section: test-abc123def456-section')
            expect(textContent).toContain(`id=${TEST_IDS.SECTION_1}`)
            expect(textContent).toContain('Use tasks-list-for-container with type=section')

            // Verify structured content
            const structuredContent = extractStructuredContent(result)
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    section: expect.objectContaining({
                        id: TEST_IDS.SECTION_1,
                        name: 'test-abc123def456-section',
                    }),
                    operation: 'created',
                }),
            )
        })

        it('should handle different section properties from API', async () => {
            const mockApiResponse = createMockSection({
                id: TEST_IDS.SECTION_2,
                projectId: 'project-789',
                sectionOrder: 2,
                name: 'My Section Name',
            })

            mockTodoistApi.addSection.mockResolvedValue(mockApiResponse)

            const result = await sectionsManage.execute(
                { name: 'My Section Name', projectId: 'project-789' },
                mockTodoistApi,
            )

            expect(mockTodoistApi.addSection).toHaveBeenCalledWith({
                name: 'My Section Name',
                projectId: 'project-789',
            })

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Created section: My Section Name')
            expect(textContent).toContain(`id=${TEST_IDS.SECTION_2}`)
            expect(textContent).toContain('Use tasks-add-multiple with sectionId')

            // Verify structured content
            const structuredContent = extractStructuredContent(result)
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    section: expect.objectContaining({
                        id: TEST_IDS.SECTION_2,
                        name: 'My Section Name',
                    }),
                    operation: 'created',
                }),
            )
        })

        it('should throw error when projectId is missing for new section', async () => {
            await expect(
                sectionsManage.execute({ name: 'test-section' }, mockTodoistApi),
            ).rejects.toThrow(
                'Error: projectId is required when creating a new section (when id is not provided).',
            )

            expect(mockTodoistApi.addSection).not.toHaveBeenCalled()
        })
    })

    describe('updating an existing section', () => {
        it('should update a section when id is provided', async () => {
            const mockApiResponse: Section = {
                id: 'existing-section-123',
                projectId: '6cfCcrrCFg2xP94Q',
                sectionOrder: 1,
                userId: 'test-user',
                addedAt: '2024-01-01T00:00:00Z',
                updatedAt: '2024-01-01T00:00:00Z',
                archivedAt: null,
                isArchived: false,
                isDeleted: false,
                isCollapsed: false,
                name: 'Updated Section Name',
            }

            mockTodoistApi.updateSection.mockResolvedValue(mockApiResponse)

            const result = await sectionsManage.execute(
                { id: 'existing-section-123', name: 'Updated Section Name' },
                mockTodoistApi,
            )

            expect(mockTodoistApi.updateSection).toHaveBeenCalledWith('existing-section-123', {
                name: 'Updated Section Name',
            })

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Updated section: Updated Section Name')
            expect(textContent).toContain('id=existing-section-123')
            expect(textContent).toContain('Use tasks-list-for-container with type=section')

            // Verify structured content
            const structuredContent = extractStructuredContent(result)
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    section: expect.objectContaining({
                        id: 'existing-section-123',
                        name: 'Updated Section Name',
                    }),
                    operation: 'updated',
                }),
            )
        })

        it('should update section without requiring projectId', async () => {
            const mockApiResponse: Section = {
                id: 'section-update-test',
                projectId: 'original-project-id',
                sectionOrder: 3,
                userId: 'test-user',
                addedAt: '2024-01-01T00:00:00Z',
                updatedAt: '2024-01-01T00:00:00Z',
                archivedAt: null,
                isArchived: false,
                isDeleted: false,
                isCollapsed: false,
                name: 'Section New Name',
            }

            mockTodoistApi.updateSection.mockResolvedValue(mockApiResponse)

            const result = await sectionsManage.execute(
                {
                    id: 'section-update-test',
                    name: 'Section New Name',
                    // Note: projectId not provided for update
                },
                mockTodoistApi,
            )

            expect(mockTodoistApi.updateSection).toHaveBeenCalledWith('section-update-test', {
                name: 'Section New Name',
            })

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Updated section: Section New Name')
            expect(textContent).toContain('id=section-update-test')
            expect(textContent).toContain('Use tasks-list-for-container with type=section')

            // Verify structured content
            const structuredContent = extractStructuredContent(result)
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    section: expect.objectContaining({
                        id: 'section-update-test',
                        name: 'Section New Name',
                    }),
                    operation: 'updated',
                }),
            )
        })
    })

    describe('error handling', () => {
        it('should propagate API errors for section creation', async () => {
            const apiError = new Error('API Error: Section name is required')
            mockTodoistApi.addSection.mockRejectedValue(apiError)

            await expect(
                sectionsManage.execute({ name: '', projectId: '6cfCcrrCFg2xP94Q' }, mockTodoistApi),
            ).rejects.toThrow('API Error: Section name is required')
        })

        it('should propagate API errors for section updates', async () => {
            const apiError = new Error('API Error: Section not found')
            mockTodoistApi.updateSection.mockRejectedValue(apiError)

            await expect(
                sectionsManage.execute(
                    { id: 'non-existent-section', name: 'Updated Name' },
                    mockTodoistApi,
                ),
            ).rejects.toThrow('API Error: Section not found')
        })
    })
})
