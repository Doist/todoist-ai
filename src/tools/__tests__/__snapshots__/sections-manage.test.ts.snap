// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`sections-manage tool creating a new section should create a section and return result 1`] = `
"Created section: test-abc123def456-section • id=section-123
Next:
- Use tasks-add-multiple with sectionId=section-123 to add your first tasks
- Use tasks-list-for-container with type=section and id=section-123 to verify setup
- Use overview with projectId=6cfCcrrCFg2xP94Q to see project organization"
`;

exports[`sections-manage tool creating a new section should handle different section properties from API 1`] = `
"Created section: My Section Name • id=section-456
Next:
- Use tasks-add-multiple with sectionId=section-456 to add your first tasks
- Use tasks-list-for-container with type=section and id=section-456 to verify setup
- Use overview with projectId=project-789 to see project organization"
`;

exports[`sections-manage tool updating an existing section should update a section when id is provided 1`] = `
"Updated section: Updated Section Name • id=existing-section-123
Next:
- Use tasks-list-for-container with type=section and id=existing-section-123 to see existing tasks
- Use sections-search to see all sections in this project
- Consider updating task descriptions if section purpose changed"
`;

exports[`sections-manage tool updating an existing section should update section without requiring projectId 1`] = `
"Updated section: Section New Name • id=section-update-test
Next:
- Use tasks-list-for-container with type=section and id=section-update-test to see existing tasks
- Use sections-search to see all sections in this project
- Consider updating task descriptions if section purpose changed"
`;
